### PR TITLE
12 Add local dev containers

### PR DESCRIPTION
**Pull request by sophie22**
This update adds a local development Docker stack, extends the flexibility of the `elastic_utils`, adds a script to load the synthetic example data into the local ElasticSearch instance and adds a script to run the legacy BRCA query on this local ElasticSearch instance.

Add a local development Docker stack including a local ElasticSearch and Kibana interface, as well as the mlflow, minIO and Doccano containers. Local ElasticSearch configuration was taken from [here](https://github.com/elastic/start-local) with default settings. Start, stop and uninstall scripts are automatically generated.

To set up the local environment, do `docker compose up -d` from within the `local_dev` folder *having updated the environment variables* in a copy of the `.env.example`.

Various packages might need to be added to the local virtual environment (might be worth adding a requirements file in the root of the repo?).
To load the synthetic data into the locally running ES, run the `synthetic_brca_load_es.py` script with completed environment variables with the containers running in an appropriate virtual environment.
To run a test query on this local dataset, execute the `breast_brca_local.py` script within the same setup. This results in output files being placed in `projects/brca_legacy/data/breast_brca_status/` folder.

### CODE REVIEW

File: local_dev/stop.sh
Comment (drjzhn) [inline local_dev/stop.sh:1]: The start script makes a lot of sense. This stop script seems a little redundant to me as it's just a wrapper around a single ```docker compose``` command? Couldn't we just run ```docker-compose stop``` / ```docker-compose down``` directly from command line.

File: local_dev/uninstall.sh
Comment (drjzhn) [inline local_dev/uninstall.sh:1]: Can we also get rid of this one as it destroys the entire dev setup (```rm docker-compose.yml .env uninstall.sh start.sh stop.sh```).

File: src/bioext/elastic_utils.py
Comment (drjzhn) [inline src/bioext/elastic_utils.py:1]: Nice. I've just changed to basic_auth as http_auth has been deprecated. See https://www.elastic.co/guide/en/elasticsearch/client/python-api/current/migration.html (meant to make a suggestion but accidently committed). Also added some suggestions for error handling for when users haven't sent environment variables properly (like I did when I was testing just now!!)

File: src/bioext/elastic_utils.py [hunk 1]
```diff
@@ -11,25 +11,38 @@ def __init__(self, *args, **kwargs) -> None:
...
-        self.proxy_node = GsttProxyNode
+        if proxy:
+            self.proxy_node = GsttProxyNode
+        else:
+            self.proxy_node = None
 
-        self.es = self.create_session()
+        if conn_mode == "API":
+            self.api_id = os.getenv("ELASTIC_API_ID")
+            self.api_key = os.getenv("ELASTIC_API_KEY")
```
Comment (drjzhn) [inline src/bioext/elastic_utils.py:26]: Adding some error handling

```suggestion
        try:
            if conn_mode == "API":
                self.api_id = os.getenv("ELASTIC_API_ID")
                self.api_key = os.getenv("ELASTIC_API_KEY")
                
                if not all([self.api_id, self.api_key]):
                    raise ValueError("Check that ELASTIC_API_ID and ELASTIC_API_KEY are in env variables")
```

File: src/bioext/elastic_utils.py [hunk 2]
```diff
@@ -11,25 +11,38 @@ def __init__(self, *args, **kwargs) -> None:
...
+        if conn_mode == "HTTP":
+            self.es_user = os.getenv("ELASTIC_USER")
+            self.es_pwd = os.getenv("ELASTIC_PWD")
+
+            self.es = Elasticsearch(
+                hosts=self.es_server,
+                basic_auth=(self.es_user, self.es_pwd),
+                verify_certs=False,
+                ssl_show_warn=False
+            )
```
Comment (drjzhn) [inline src/bioext/elastic_utils.py:45]: And something like this

```suggestion
                self.es = Elasticsearch(
                    hosts=self.es_server,
                    api_key=(self.api_id, self.api_key),
                    node_class=self.proxy_node,
                    verify_certs=False,
                    ssl_show_warn=False
                )

            elif conn_mode == "HTTP":
                self.es_user = os.getenv("ELASTIC_USER")
                self.es_pwd = os.getenv("ELASTIC_PWD")
                
                if not all([self.es_user, self.es_pwd]):
                    raise ValueError("Check ELASTIC_USER and ELASTIC_PWD are in env variables")
                        
                self.es = Elasticsearch(
                    hosts=self.es_server,
                    basic_auth=(self.es_user, self.es_pwd),  # http_auth has been deprecated
                    verify_certs=False,
                    ssl_show_warn=False
                )
                
            else:
                raise ValueError("Argument conn_mode must be 'HTTP' or 'API'")    
```

File: projects/brca_legacy/breast_brca_local.py
Comment (drjzhn) [inline projects/brca_legacy/breast_brca_local.py:1]: Could we have a dedicated folder for local projects, and keep each set of related scripts in its own project folder. E.g. ```projects/local/synthetic_brca```. Let's discuss, or I can show you what I mean over a call.

File: projects/synthetic_brca/synthetic_brca_load_es.py
Comment (drjzhn) [inline projects/synthetic_brca/synthetic_brca_load_es.py:1]: See comment above about a dedicated local folder, and projects/local/specific_project subfolders.

**Review by drjzhn — COMMENTED** [review]
Great work Sophie, thanks for this! Small suggestions. Easier to chat, but in general:

(1) Because the repo is setup as a package, we dont need to install elastic/doccano etc separately, as they are installed as part of the package dependencies (```-e ../..``` in requirements.txt). So only extra packages need to go into the requirements.txt for each project, that are specific to that project.

(2) Could add ELASTIC_USER=elastic, ELASTIC_PWD, and ELASTIC_SERVER variables to the env.example file in whereever we put the local synthetic data project

(3) Let's discuss how to document in a call. 

(4) Along those lines, I've been lazy about docstrings and type hinting etc...we can introduce minimum standards for these and I'll rectify.

File: src/bioext/elastic_utils.py (hunk 1 above)
Comment (lawrenceadams) [reply src/bioext/elastic_utils.py:26]: Could also just use `os.environ[]` which will KeyError if it doesn't exist

File: local_dev/stop.sh
Comment (sophie22) [reply local_dev/stop.sh:1]: yes, we can use docker commands directly.

File: projects/brca_legacy/breast_brca_local.py
Comment (sophie22) [reply projects/brca_legacy/breast_brca_local.py:1]: added new local project folder

**Comment by sophie22** [comment]
Minor updates and corrections to the repo READme and the elastic_utils.
Rearranged local deployment and example project files into new folders as discussed, including a description for how to use and test the local environment.

**Review by drjzhn — APPROVED** [review]
Nice work. The local dev is a great idea - I will be using it also!

**Pull request merged**